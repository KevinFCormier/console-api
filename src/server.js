/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import https from 'https';
import fs from 'fs';
import schema from './schema/index';
import config from '../config';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const GRAPHQL_PORT = process.env.PORT || config.get('httpPort') || 4000;
const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(req => ({ formatError, schema, context: req })));

const privateKey = fs.readFileSync(process.env.serverKey || './sslcert/hcmuiapi.key', 'utf8');
const certificate = fs.readFileSync(process.env.serverCert || './sslcert/hcmuiapi.crt', 'utf8');
const credentials = { key: privateKey, cert: certificate };
const server = https.createServer(credentials, graphQLServer);

server.listen(GRAPHQL_PORT, () => {
  logger.info(`[pid ${process.pid}] [env ${process.env.NODE_ENV}] started.`);
  logger.info(`HCM UI API is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphql`);
  if (process.env.NODE_ENV !== 'production') {
    logger.info(`GraphiQL is now running on https://localhost:${GRAPHQL_PORT}${CONTEXT_PATH}/graphiql`);
  }
});
