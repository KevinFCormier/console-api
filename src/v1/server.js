/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import express from 'express';
import { graphqlExpress, graphiqlExpress } from 'apollo-server-express';
import { isInstance as isApolloErrorInstance, formatError as formatApolloError } from 'apollo-errors';
import bodyParser from 'body-parser';
import { app as inspect } from '@icp/security-middleware';
import log4js from 'log4js';
import morgan from 'morgan';
import helmet from 'helmet';
import cookieParser from 'cookie-parser';
import schema from './schema/index';
import config from '../../config';
import HCMConnector from './datasource/lib/hcm-connector';
import MongoConnector from './datasource/lib/mongo-connector';

const logger = log4js.getLogger('server');

const log4jsConfig = process.env.LOG4JS_CONFIG ? JSON.parse(process.env.LOG4JS_CONFIG) : undefined;
log4js.configure(log4jsConfig || 'config/log4js.json');

const CONTEXT_PATH = config.get('contextPath');

const graphQLServer = express();
graphQLServer.use('*', helmet());

if (process.env.NODE_ENV === 'production') {
  graphQLServer.use(
    '*',
    morgan('combined', {
      skip: (req, res) => res.statusCode < 400,
    }),
  );
  logger.info('Authentication enabled');
  graphQLServer.use(cookieParser(), inspect);
} else {
  graphQLServer.use('*', morgan('dev'));
  // disable security check and enable graphiql only for local dev
  graphQLServer.use(cookieParser());
  graphQLServer.use(`${CONTEXT_PATH}/graphiql`, graphiqlExpress({ endpointURL: `${CONTEXT_PATH}/graphql` }));
}

const formatError = (error) => {
  const { originalError } = error;
  if (isApolloErrorInstance(originalError)) {
    logger.error(JSON.stringify(error.originalError, null, 2));
  }
  return formatApolloError(error);
};


const getProductionOptions = async (req) => {
  const mongoConnector = new MongoConnector();
  const hcmConnector = new HCMConnector({
    hcmUrl: config.get('hcmUrl') || 'http://localhost:8080',
    pollInterval: config.get('hcmPollInterval') || 200,
    pollTimeout: config.get('hcmPollTimeout') || 10000,
  });

  await mongoConnector.connect(config.get('mongodbUrl') || 'mongodb://localhost:27017/weave');
  return {
    formatError,
    schema,
    context: {
      req,
      hcmConnector,
      mongoConnector,
    },
  };
};

graphQLServer.use(`${CONTEXT_PATH}/graphql`, bodyParser.json(), graphqlExpress(getProductionOptions));

export default graphQLServer;
