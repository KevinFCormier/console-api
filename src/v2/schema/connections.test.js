/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Cloud Connections', () => {
  test('Correctly Resolves Get Cloud Connections', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
            {
              connections {
                 metadata{
                    name
                    namespace
                    provider
                    name_namespace
                 }
              }
            }`,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Get Cloud Providers', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
            {
              providers {
                name
                longname
                configMetadata
                configValues
                clusterMetadata
                clusterValues
                statusCode
              }
            }`,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Create Cloud Connection', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
            mutation {
                createCloudConnection(body: {name: "iks-test", namespace: "default", provider: "iks", metadata: "secret:\n  apiKey: iks-test\n"})
            }
          `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Edit Cloud Connection', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
            mutation {
                editCloudConnection(body: {name: "iks-test", namespace: "default", provider: "iks", metadata: "secret:\n  apiKey: iks-test\n"}, namespace:"default", name:"iks-test")
            }
          `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Delete Cloud Connection', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
            mutation {
                deleteCloudConnection(namespace:"default", name:"iks-test")
            }
          `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});
