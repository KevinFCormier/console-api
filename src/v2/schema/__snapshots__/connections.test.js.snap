// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloud Connections Correctly Resolves Create Cloud Connection 1`] = `
Object {
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 121,
          "line": 3,
        },
      ],
      "message": "Syntax Error: Unterminated string.",
    },
  ],
}
`;

exports[`Cloud Connections Correctly Resolves Delete Cloud Connection 1`] = `
Object {
  "data": Object {
    "deleteCloudConnection": null,
  },
}
`;

exports[`Cloud Connections Correctly Resolves Edit Cloud Connection 1`] = `
Object {
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 119,
          "line": 3,
        },
      ],
      "message": "Syntax Error: Unterminated string.",
    },
  ],
}
`;

exports[`Cloud Connections Correctly Resolves Get Cloud Connections 1`] = `
Object {
  "data": Object {
    "connections": Array [
      Object {
        "metadata": Object {
          "name": null,
          "name_namespace": null,
          "namespace": null,
          "provider": null,
        },
      },
    ],
  },
}
`;

exports[`Cloud Connections Correctly Resolves Get Cloud Providers 1`] = `
Object {
  "data": Object {
    "providers": Array [
      Object {
        "clusterMetadata": "",
        "clusterValues": "",
        "configMetadata": "",
        "configValues": "",
        "longname": "iks-test",
        "name": "iks-test",
        "statusCode": 200,
      },
    ],
  },
}
`;
